### PR TITLE
[Merged by Bors] - fead: Add `lift_unique`, `lift_unique'` to `IsFractionRing`

### PR DESCRIPTION
**Pull request by yhtq**
Add `lift_unique`, `lift_unique'` to `IsFractionRing`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Localization/FractionRing.lean [hunk 1]
```diff
@@ -222,6 +222,23 @@ such that `z = f x * (f y)⁻¹`. -/
...
+  IsLocalization.lift_unique _ hf1
+
+/-- Another version of unique to give two lift maps should be equal -/
+theorem lift_unique'
+    {L: Type*} [Field L]
+    {g: A →+* L} (hg: Function.Injective g)
+    {f1 f2: K →+* L}
+    (hf1: ∀ x, f1 (algebraMap A K x) = g x)
+    (hf2: ∀ x, f2 (algebraMap A K x) = g x): f1 = f2 :=
+  Eq.trans (lift_unique hg hf1).symm (lift_unique hg hf2)
```
Comment (acmepjz) [inline Mathlib/RingTheory/Localization/FractionRing.lean:239]: Is this really useful? I mean `rw [lift_unique' hg hf1 hf2]` can be just `rw [← lift_unique hg hf1, lift_unique hg hf2]`.

Comment (yhtq) [reply Mathlib/RingTheory/Localization/FractionRing.lean:239]: > Is this really useful? I mean `rw [lift_unique' hg hf1 hf2]` can be just `rw [← lift_unique hg hf1, lift_unique hg hf2]`.

For example. if the goal is `f1.comp f2 = g1.comp g2`, you can direcly `apply lift_unique'` and then prove the conditions, but `rw` may not find the correct function (unless we prove the conditions first and pass to it).

Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FractionRing.lean:239]: OK.

File: Mathlib/RingTheory/Localization/FractionRing.lean [hunk 2]
```diff
@@ -222,6 +222,23 @@ such that `z = f x * (f y)⁻¹`. -/
 noncomputable def lift (hg : Injective g) : K →+* L :=
   IsLocalization.lift fun y : nonZeroDivisors A => isUnit_map_of_injective hg y
 
+theorem lift_unique
+    {L: Type*} [Field L]
```
[the remark this answers is not in the document]
Comment (yhtq) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: Sorry for missing that

File: Mathlib/RingTheory/Localization/FractionRing.lean [hunk 3]
```diff
@@ -222,6 +222,23 @@ such that `z = f x * (f y)⁻¹`. -/
...
+  IsLocalization.lift_unique _ hf1
+
+/-- Another version of unique to give two lift maps should be equal -/
+theorem lift_unique'
+    {L : Type*} [Field L]
+    {g : A →+* L} (hg : Function.Injective g)
+    {f1 f2 : K →+* L}
+    (hf1 : ∀ x, f1 (algebraMap A K x) = g x)
+    (hf2 : ∀ x, f2 (algebraMap A K x) = g x) : f1 = f2 :=
+  Eq.trans (lift_unique hg hf1).symm (lift_unique hg hf2)
```
[the remark this answers is not in the document]
Comment (yhtq) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: I have not thought of that it dosesn't depend on `g`. Thanks!

File: Mathlib/RingTheory/Localization/FractionRing.lean [hunk 4]
```diff
@@ -222,12 +222,15 @@ such that `z = f x * (f y)⁻¹`. -/
...
+theorem lift_unique
+    (hg : Function.Injective g)
+    {f : K →+* L}
     (hf1 : ∀ x, f (algebraMap A K x) = g x) : IsFractionRing.lift hg = f :=
   IsLocalization.lift_unique _ hf1
 
 /-- Another version of unique to give two lift maps should be equal -/
-theorem lift_unique' {f1 f2 : K →+* L}
+theorem lift_unique'
+    {f1 f2 : K →+* L}
```
Comment (tb65536) [inline Mathlib/RingTheory/Localization/FractionRing.lean:233]: I don't think putting these variables on their own lines is the preferred style. For instance, see how some of the theorems on this page are structured: https://leanprover-community.github.io/contribute/style.html

Comment (yhtq) [reply Mathlib/RingTheory/Localization/FractionRing.lean:233]: emm I think you're right. Sorry.

File: Mathlib/RingTheory/Localization/FractionRing.lean (hunk 3 above)
[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: I think this statement is very similar to that of [IsPurelyInseparable.injective_comp_algebraMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#IsPurelyInseparable.injective_comp_algebraMap). Do you think it's a good idea to rephrase this as `Function.Injective fun (f : K →+* L) => f.comp (algebraMap A K)` and rename the result as `IsFractionRing.injective_comp_algebraMap`?

[the remark this answers is not in the document]
Comment (yhtq) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: > I think this statement is very similar to that of [IsPurelyInseparable.injective_comp_algebraMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/PurelyInseparable.html#IsPurelyInseparable.injective_comp_algebraMap). Do you think it's a good idea to rephrase this as `Function.Injective fun (f : K →+* L) => f.comp (algebraMap A K)` and rename the result as `IsFractionRing.injective_comp_algebraMap`?

It's a good statement but the intuition is slightly different from that of the direct form I think? Maybe I can implement both of them?

[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: I think it's the same as your `lift_unique'` besides that the condition is `RingHom.comp f1 (algebraMap A K) = RingHom.comp f2 (algebraMap A K)`. It's the same after `ext`.

[the remark this answers is not in the document]
Comment (yhtq) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: Of course they are same, but I mean 

> I think it's the same as your `lift_unique'` besides that the condition is `RingHom.comp f1 (algebraMap A K) = RingHom.comp f2 (algebraMap A K)`. It's the same after `ext`.

Of course they are the same. I mean that in some circumstances one may only consider "two homomorphisms are equal" rather than "the derived map is injective" but in other circumstances it is opposite.

[the remark this answers is not in the document]
Comment (acmepjz) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: > in some circumstances one may only consider "two homomorphisms are equal" rather than "the derived map is injective"

I mean what `lift_unique'` can do can also be done by `injective_comp_algebraMap`. For example, `apply lift_unique'` can give `∀ x : A, f1 (algebraMap A K x) = f2 (algebraMap A K x)`, on the other hand, `apply injective_comp_algebraMap; ext` can also give `∀ x : A, f1 (algebraMap A K x) = f2 (algebraMap A K x)`.

[the remark this answers is not in the document]
Comment (yhtq) [reply Mathlib/RingTheory/Localization/FractionRing.lean]: > > in some circumstances one may only consider "two homomorphisms are equal" rather than "the derived map is injective"
> 
> I mean what `lift_unique'` can do can also be done by `injective_comp_algebraMap`. For example, `apply lift_unique'` can give `∀ x : A, f1 (algebraMap A K x) = f2 (algebraMap A K x)`, on the other hand, `apply injective_comp_algebraMap; ext` can also give `∀ x : A, f1 (algebraMap A K x) = f2 (algebraMap A K x)`.

I know they are the same in Lean. I just means they have different intuitions (in mind) so implement both versions may be better, but of course it's not a big deal(:

File: Mathlib/RingTheory/Localization/FractionRing.lean [hunk 5]
```diff
@@ -222,6 +222,18 @@ such that `z = f x * (f y)⁻¹`. -/
 noncomputable def lift (hg : Injective g) : K →+* L :=
   IsLocalization.lift fun y : nonZeroDivisors A => isUnit_map_of_injective hg y
 
+theorem lift_unique (hg : Function.Injective g) {f : K →+* L}
+    (hf1 : ∀ x, f (algebraMap A K x) = g x) : IsFractionRing.lift hg = f :=
+  IsLocalization.lift_unique _ hf1
+
+/-- Another version of unique to give two lift maps should be equal -/
+theorem lift_unique' {f1 f2 : K →+* L}
```
Comment (jcommelin) [inline Mathlib/RingTheory/Localization/FractionRing.lean:230]:
```suggestion
theorem ringHom_ext {f1 f2 : K →+* L}
```
I will leave it up to you to decide if you want to keep the docstring as is.

**Comment by jcommelin** [comment]
CI isn't happy.

File: Mathlib/RingTheory/Localization/FractionRing.lean [hunk 6]
```diff
@@ -222,6 +222,21 @@ such that `z = f x * (f y)⁻¹`. -/
...
+/-- Another version of unique to give two lift maps should be equal -/
+theorem ringHom_ext {f1 f2 : K →+* L}
+    (hf : ∀ x : A, f1 (algebraMap A K x) = f2 (algebraMap A K x)) : f1 = f2 := by
+  ext z
+  obtain ⟨x, y, hy, rfl⟩ := IsFractionRing.div_surjective (A := A) z
+  rw [map_div₀, map_div₀, hf, hf]
+
+theorem injective_comp_algebraMap
+    : Function.Injective fun (f : K →+* L) => f.comp (algebraMap A K) :=
+  fun _ _ h => lift_unique' (fun x => RingHom.congr_fun h x)
```
Comment (acmepjz) [inline Mathlib/RingTheory/Localization/FractionRing.lean:238]:
```suggestion
  fun _ _ h => ringHom_ext (fun x => RingHom.congr_fun h x)
```

File: Mathlib/RingTheory/Localization/FractionRing.lean [hunk 7]
```diff
@@ -222,6 +222,21 @@ such that `z = f x * (f y)⁻¹`. -/
...
+
+/-- Another version of unique to give two lift maps should be equal -/
+theorem ringHom_ext {f1 f2 : K →+* L}
+    (hf : ∀ x : A, f1 (algebraMap A K x) = f2 (algebraMap A K x)) : f1 = f2 := by
+  ext z
+  obtain ⟨x, y, hy, rfl⟩ := IsFractionRing.div_surjective (A := A) z
+  rw [map_div₀, map_div₀, hf, hf]
+
+theorem injective_comp_algebraMap
+    : Function.Injective fun (f : K →+* L) => f.comp (algebraMap A K) :=
```
Comment (acmepjz) [inline Mathlib/RingTheory/Localization/FractionRing.lean:237]:
```suggestion
theorem injective_comp_algebraMap :
    Function.Injective fun (f : K →+* L) => f.comp (algebraMap A K) :=
```

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge
